tests/test_data/test_filter.py

```diff
@@ -1,9 +1,555 @@
+        )
+        expected_df = pl.DataFrame(
+            {
+                "ADDRESS_DETAIL_PID": ["1003"],
+                "LATITUDE": [36.0],
+                "LONGITUDE": [148.5],
+                "FLAT_TYPE_CODE": ["unit"],
+                "POSTCODE": [3000],
+            }
+        )
+
+        result_lf = filter_and_join_gnaf_frames(
+            default_geocode_data,
+            address_detail_data,
+            building_types=["unit"],
+            postcodes=[3000],
+        )
+        expected_df = pl.DataFrame(
+            {
+                "ADDRESS_DETAIL_PID": ["1003"],
+                "LATITUDE": [36.0],
+                "LONGITUDE": [148.5],
+                "FLAT_TYPE_CODE": ["unit"],
+                "POSTCODE": [3000],
+            }
+        )
+
+        assert result_lf.collect().to_dicts() == expected_df.to_dicts()
+
+    def test_no_matching_building_type(self, default_geocode_data, address_detail_data):
+        """
+        Test the function when no rows match the building type filter.
+        """
+        result_lf = filter_and_join_gnaf_frames(
+            default_geocode_data, address_detail_data, building_types=["apartment"]
+        )
+        expected_df = pl.DataFrame(
+            {
+                "ADDRESS_DETAIL_PID": [],
+                "LATITUDE": [],
+                "LONGITUDE": [],
+                "FLAT_TYPE_CODE": [],
+                "POSTCODE": [],
+            }
+        )
+
+        result_lf = filter_and_join_gnaf_frames(
+            default_geocode_data, address_detail_data, building_types=["apartment"]
+        )
+        expected_df = pl.DataFrame(
+            {
+                "ADDRESS_DETAIL_PID": [],
+                "LATITUDE": [],
+                "LONGITUDE": [],
+                "FLAT_TYPE_CODE": [],
+                "POSTCODE": [],
+            }
+        )
+
+        assert result_lf.collect().to_dicts() == expected_df.to_dicts()
+
+    def test_no_matching_postcode(self, default_geocode_data, address_detail_data):
+        """
+        Test the function when no rows match the postcode filter.
+        """
+        result_lf = filter_and_join_gnaf_frames(
+            default_geocode_data, address_detail_data, postcodes=[9999]
+        )
+        expected_df = pl.DataFrame(
+            {
+                "ADDRESS_DETAIL_PID": [],
+                "LATITUDE": [],
+                "LONGITUDE": [],
+                "FLAT_TYPE_CODE": [],
+                "POSTCODE": [],
+            }
+        )
+
+        result_lf = filter_and_join_gnaf_frames(
+            default_geocode_data, address_detail_data, postcodes=[9999]
+        )
+        expected_df = pl.DataFrame(
+            {
+                "ADDRESS_DETAIL_PID": [],
+                "LATITUDE": [],
+                "LONGITUDE": [],
+                "FLAT_TYPE_CODE": [],
+                "POSTCODE": [],
+            }
+        )
+
+        assert result_lf.collect().to_dicts() == expected_df.to_dicts()
```

Comment:
why do you have two test classes? some test methods also have redundant code e.g. code after a return statement, please clean this up

Reply:
btw I've changed the docstring, these test cases are the only thing need fixing before merging

Reply:
Sorry about the duplicate test cases, my bad! Fixed it now.